File: UnitConversionLibrary/Java/UnitConversion/ConstantGroup.java
/////////////////////////////////////////////////////////////////////////////////
//2345678901234567890123456789012345678901234567890123456789012345678901234567890
/////////////////////////////////////////////////////////////////////////////////
// PROJECT: Unit Conversion
//
// Copyright Copyright 2024 MAP
//
// Unpublished - Rights reserved under the Copyright Laws of the United States
//
/////////////////////////////////////////////////////////////////////////////////
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful, 
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
//
/////////////////////////////////////////////////////////////////////////////////
//
// File ConstantGroup.java
//
// ConstantGroup is the base class for derived units, physical constants, 
// physical  definition, physical measurements, WGS84 Model,  and some 
// SI units. These  'units'  are used as the basis for other units but are not 
// directly convertible as such. They are stored as UBASE objects, as other units 
// are, and placed in a map and are accessible for reference or other purposes.     
//
//Version: 1.0
//
/////////////////////////////////////////////////////////////////////////////////

package UnitConversion;

import java.util.*;

public class ConstantGroup
{
   /*
    * Default constructor.
    */
    public ConstantGroup()
    {
        _map = new HashMap<String, UBASE>();
        _name = "Invalid";
        _valid = false;
        _version = "Invalid";
    }

   /*
    * Constructor.
    *
    * @param name    - (input) name of the constant group.
    * @param version - (input) software version.
    */
    public ConstantGroup(String name, 
                         String version)
    {
        _map = new HashMap<String, UBASE>();
        _name = name;
        _valid = true;
        _version = version;
    }

   /*
    * Constructor.
    *
    * @param name      - (input) name of the constant group.
    * @param version   - (input) software version.
    * @param constants - (input) constants in the group.
    */
    public ConstantGroup(String name, 
                         String version, 
                         HashMap<String, UBASE> constants)
    {
        _map = copyMap(constants);
        _name = name;
        _valid = true;
        _version = version;
    }

   /*
    * Copy constructor
    *
    * @param other - (input) the other ConstantGroup to be copied.
    */
    public ConstantGroup(ConstantGroup other)
    {
        _map = new HashMap<String, UBASE>();
        for(Map.Entry<String, UBASE> entry : other._map.entrySet())
        {
            _map.put(entry.getKey(),new UBASE(entry.getValue()));
        }
        _name = other._name;
        _valid = other._valid;
        _version = other._version;
    }

   /*
    * The check method checks if the software version matches with the unit
    * versions of the units in the ConstantGroup. Units are auto generated from 
    * an Excel spreadsheet using VBA code.
    * 
    * @return - true if software and unit versions match, false otherwise.
    */
    public boolean check()
    {
        Version v = Version.Instance();
        for(Map.Entry<String, UBASE> entry : _map.entrySet())
        {
            UBASE db = entry.getValue();
            if (db.version() == v.version())
            {
                continue;
            }
            else
            {
                return false;
            }
        }
        return true;
    }
    
   /*
    * Get named constant.
    *
    * @param name - (input) name of constant to get.
    *
    * @return - the specified constant or invalid constant if not found.
    */
    public UBASE constant(String name)
    {
       if (_map.containsKey(name))
       {
           return _map.get(name);
                
       }
       else
       {
           return new UBASE();
       }
    }

   /*
    * Get list of constant names.
    *
    * @return - a list of constant names.
    */
     public List<String> constantNames()
    {
        return new ArrayList<String>(_map.keySet());
    }

   /*
    * Get the name of the ConstantGroup.
    *
    * @return - the name of the ConstantGroup.
    */
    public String name()
    {
        return _name;
    }

    /* 
     * Check if input is error code..
     *
     * @param x       - (input) value to check.
     *
     * @return - true if input is error code, false otherwise.
     */
    public boolean isError(double x)
    {
        return UBASE.isERROR(x);
    }

   /*
    * Get list of system names.
    *
    * @return - a list of system names.
    */
    public List<String> systemNames()
    {
        ArrayList<String> sys = new ArrayList<String>();
        for(Map.Entry<String, UBASE> entry : _map.entrySet())
        {
            UBASE db = entry.getValue();
            if(sys.contains(db.system()))
            {
               continue;
            }
            else
            {
                sys.add(db.system());
            }
        }
        return sys;
    }

   /*
    * Get list of constant types.
    *
    * @return - a list of constant types.
    */
     public List<String> typeNames()
    {
        ArrayList<String> al = new ArrayList<String>();
        al.add(_name);
        return al;
    }

   /*
    * Get the validity of the ConstantGroup.
    *
    * @return - true if ConstantGroup is valid, false otherwise.
    */
    public boolean valid()
    {
        return _valid;
    }
        
    /*
     * Get the value of a constant.
     *
     * @param constName - (input) the name of the constant.
     *
     * @return the value of the named constant.
     */
    public double value(String constName)
    {
         return constant(constName).value().asDouble();
    }

    /*
     * Get the software version.
     *
     * @return  - the software version.
     */
    public String version()
    {
        return _version;
    }

   /*
    * Create a deep copy of a HashMap.
    * 
    * @param - map - (input) the map to create a deep copy of.
    *
    * @return - a deep copy of the input map.
    */
    private HashMap<String, UBASE> copyMap(HashMap<String, UBASE> map)
    {
        HashMap<String, UBASE> cpyMap = new HashMap<String, UBASE>();
        for(Map.Entry<String, UBASE> entry : map.entrySet())
        {
            cpyMap.put(entry.getKey(),new UBASE(entry.getValue()));
        }
        return cpyMap;
    }
      
    //**************************************************************************
    // Class data members
    //**************************************************************************
    /* 
     * ConstantGroup is the base class for derived unit, physical constants,  
     * physical definitions.physical measurements, WGS84 Model,  and SI units.  
     * These 'units' are used as the basis for other units but are not directly 
     * convertible as such. They are stored as UBASE objects, as other units 
     * are, and placed in a HashMap and are accessible for reference or other 
     * purposes.     
     */
    protected HashMap<String, UBASE> _map;
    
    /*
     * Name for the class of constants stored.
     */
    protected String _name;
    
    /*
     * Flag that is set to true if ConstantGroup contains one or more valid 
     * constants.
     */
    protected boolean _valid;
    
    /*
     * The software version.
     */
    protected String _version;
}
// EOF

File: UnitConversionLibrary/CS/UnitConversion/ConstantGroup.cs
/////////////////////////////////////////////////////////////////////////////////
//2345678901234567890123456789012345678901234567890123456789012345678901234567890
/////////////////////////////////////////////////////////////////////////////////
// PROJECT: Unit Conversion
//
// Copyright Copyright 2024 MAP
//
// Unpublished - Rights reserved under the Copyright Laws of the United States
//
/////////////////////////////////////////////////////////////////////////////////
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful, 
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
//
/////////////////////////////////////////////////////////////////////////////////
//
// File ConstantGroup.cs
//
// ConstantGroup is the base class for derived units, physical constants, 
// physical  definition, physical measurements, WGS84 Model,  and some 
// SI units. These  'units'  are used as the basis for other units but are not 
// directly convertible as such. They are stored as UBASE objects, as other units 
// are, and placed in a map and are accessible for reference or other purposes.     
//
// Version: 1.0
//
/////////////////////////////////////////////////////////////////////////////////

namespace UnitConversion
{
    using System.Collections.Generic;

    /// <summary>
    /// Base class for derived units and constants. Base units are used as the 
    /// basis for other units but are not directly convertible as such.
    /// </summary>
    public class ConstantGroup
    {
        /// <value>
        /// ConstantGroup is the base class for derived units, physical  
        /// constants, physical definitions.physical measurements, WGS84 Model,  
        /// and SI units. These 'units'  are used as the basis for other units 
        /// but are not directly convertible as such.  They are stored as UBASE 
        /// objects, as other units are, and placed in a map andare accessible
        /// for reference or other purposes.     
        /// </value>
        protected Dictionary<string, UBASE> _map;

        /// <value>
        /// Name for the class of constants stored.
        /// </value>
        protected string _name;

        /// <value>
        /// Flag that is set to true if ConstantGroup contains one or more 
        /// valid constants.
        /// </value>
        protected bool _valid;

        /// <value>
        /// Software version.
        /// </value>
        protected string _version;

        /// <summary>
        /// Default constructor.
        /// </summary>
        public ConstantGroup()
        {
            _name    = "Invalid";
            _valid   = false;
            _map     = new Dictionary<string, UBASE>();
            _version = "Invalid";
        }

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param><c>name</c>      (input) name of the constant group.</param>
        /// <param><c>version</c>   (input) software version.</param>
        /// <param><c>constants</c> (input) constants in the group.</param>
        public ConstantGroup(string name, 
                             string version, 
                             Dictionary<string, UBASE> constants)
        {
            _map     = new Dictionary<string, UBASE>(constants);
            _name    = name;
            _valid   = true;
            _version = version;
        }

        /// <summary>
        /// Constructor.
        /// </summary>
        /// <param><c>name</c>    (input) name of the constant group.</param>
        /// <param><c>version</c> (input) software version.</param>
        public ConstantGroup(string name, 
                             string version)
        {
            _map     = new Dictionary<string, UBASE>();
            _name    = name;
            _valid   = true;
            _version = version;
        }

        /// <summary>
        /// Copy constructor.
        /// </summary>
        /// <param><c>other</c> (input) name of other ConstantGroup 
        ///                             to be copied.</param>
        public ConstantGroup(ConstantGroup other)
        {
            _name    = other._name;
            _valid   = other._valid;
            _version = other._version;
            _map     = new Dictionary<string, UBASE>();
            foreach (KeyValuePair<string, UBASE> entry in other._map)
            {
                _map.Add(entry.Key, new UBASE(entry.Value));
            }
        }

        /// <summary>
        /// The check method checks if the software version matches with the 
        /// unit versions of the constants in the ConstantGroup. Constants are 
        /// auto  generated from  an Excel spreadsheet using VBA code.
        /// </summary>
        /// <returns>
        /// True if software and constant versions match, false otherwise.
        /// </returns>
        public bool check()
        {
            Version v = Version.Instance();
            foreach (KeyValuePair<string, UBASE> kvp in _map)
            {
                UBASE db = kvp.Value;
                if (db.version() == v.version())
                {
                    continue;
                }
                else
                {
                    return false;
                }
            }
            return true;
        }

        /// <summary>
        /// Get named constant.
        /// </summary>
        /// <param><c>name</c> (input) name of constant to get.</param>
        /// <returns>
        /// The specified constant or invalid constant if not found.
        /// </returns>
        public UBASE constant(string name)
        {
            if (_map.ContainsKey(name))
            {
                return _map[name];

            }
            else
            {
                return new UBASE();
            }
        }

        /// <summary>
        /// Get list of constant names.
        /// </summary>
        /// <returns>
        /// A list of constant names.
        /// </returns>
        public List<string> constantNames()
        {
            return new List<string>(this._map.Keys);
        }

        /// <summary>
        /// Check if double is ERROR code.
        /// </summary>
        /// <param><c>x</c>    (input) value to check.</param>
        /// <returns>
        /// True if input is error code, false otherwise.
        /// </returns>
        public bool isError(double x)
        {
            return UBASE.isERROR(x);
        }

        /// <summary>
        /// Get the name of the ConstantGroup.
        /// </summary>
        /// <returns>
        /// The name of the ConstantGroup.
        /// </returns>
        public string name()
        {
            return _name;
        }

        /// <summary>
        /// Get list of system names.
        /// </summary>
        /// <returns>
        /// A list of system names.
        /// </returns>
        public List<string> systemNames()
        {
            List<string> sys = new List<string>();
            foreach (KeyValuePair<string, UBASE> kvp in _map)
            {
                UBASE db = kvp.Value;
                if(sys.Contains(db.system()))
                {
                    continue;
                }
                else
                {
                    sys.Add(db.system());
                }
            }
            return sys;
        }

        /// <summary>
        /// Get list of constant types.
        /// </summary>
        /// <returns>
        /// A list of constant types.
        /// </returns>
        public List<string> typeNames()
        {
            List<string> lst = new List<string>();
            lst.Add(_name);
            return lst;
        }

        /// <summary>
        /// Get the validity of the ConstantGroup.
        /// </summary>
        /// <returns>
        /// True if ConstantGroup is valid, false otherwise.
        /// </returns>
        public bool valid()
        {
            return _valid;
        }

        /// <summary>
        /// Get the value of a constant.
        /// </summary>
        /// <param><c>constName</c> (input) name of constant.</param>
        /// <returns>
        /// The value of the named constant.
        /// </returns>
        public double value(string constName)
        {
            return constant(constName).value().asDouble();
        }

        /// <summary>
        /// Get the software version.
        /// </summary>
        /// <returns>
        /// The software version.
        /// </returns>
        public string version()
        {
            return _version;
        }
    }
}
// EOF
